plugins/woocommerce/tests/e2e-pw/tests/merchant/product-delete.spec.js

```diff
@@ -0,0 +1,121 @@
+const {test: baseTest, expect} = require('@playwright/test');
+const wcApi = require('@woocommerce/woocommerce-rest-api').default;
+
+baseTest.describe('Products > Delete Product', () => {
+	baseTest.use({storageState: process.env.ADMINSTATE});
+
+	const apiFixture = baseTest.extend({
+		api: async ({baseURL}, use) => {
+			const api = new wcApi({
+				url: baseURL,
+				consumerKey: process.env.CONSUMER_KEY,
+				consumerSecret: process.env.CONSUMER_SECRET,
+				version: 'wc/v3',
+				axiosConfig: {
+					// allow 404s, so we can check if the product was deleted without try/catch
+					validateStatus: function (status) {
+						return status >= 200 && status < 300 || status === 404;
+					},
+				},
+			});
+
+			await use(api);
+		},
+	});
+
+	const test = apiFixture.extend({
+		productId: async ({baseURL, api}, use) => {
+			let productId;
+
+			await api
+				.post('products', {
+					name: 'Product to delete',
+					type: 'simple',
+					regular_price: '12.99',
+				})
+				.then((response) => {
+					productId = response.data.id;
+				});
+
+			await use(productId);
+
+			// permanently delete the product if it still exists
+			const r = await api.get(`products/${productId}`);
+			if (r.status !== 404) {
+				await api.delete(`products/${productId}`, {
+					force: true,
+				});
+			}
+		},
+	});
+
+	test('can delete a product from edit view', async ({page, productId, api}) => {
+		await test.step('Navigate to product edit page', async () => {
+			await page.goto(`wp-admin/post.php?post=${productId}&action=edit`);
+		});
+
+		await test.step('Move product to trash', async () => {
+			await page.locator('#delete-action > a').click();
+		});
+
+		await test.step('Verify product was trashed', async () => {
+			await expect(
+				page.locator('#message').last()
+			).toContainText('1 product moved to the Trash.');
+
+			const response = await api.get(`products/${productId}`);
+			await expect(response.data.status).toBe('trash');
+		});
+	});
+
+	test('can quick delete a product from product list', async ({page, productId, api}) => {
+		await test.step('Navigate to products list page', async () => {
+			await page.goto(`wp-admin/edit.php?post_type=product`);
+		});
+
+		await test.step('Move product to trash', async () => {
+			// mouse over the product row to display the quick actions
+			await page.locator(`#post-${productId}`).hover();
+
+			// move product to trash
+			await page.locator(`#post-${productId} .submitdelete`).click();
+		});
+
+		await test.step('Verify product was trashed', async () => {
+			await expect(
+				page.locator('#message').last()
+			).toContainText('1 product moved to the Trash.');
+
+			const response = await api.get(`products/${productId}`);
```

Comment:
I am not really sure we want to use API to test something in E2E tests, we have a separate API for this (I think).
Instead, can we somehow verify via UI that the product is not on the list and is indeed moved to the trash?

Reply:
> I am not really sure we want to use API to test something in E2E tests

I am curious on why is that?

> we have a separate API for this (I think)

Can you point me to it?

> Instead, can we somehow verify via UI that the product is not on the list and is indeed moved to the trash?

For sure, I can come up with some UI checks for these tests. I would avoid checking that something is not in a list, as it can quickly become a false negative. But maybe navigating to the product page and verify the displayed message seems reasonable to me. I'll work on that.

Reply:
> I am curious on why is that?
> Can you point me to it?

We have separate API tests, here is [the link](https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce/tests/api-core-tests/tests).

I don't see we have currently API test coverage for product deletion but there are similar scenarios like deleting orders. I intended that it might have been better to use visual checks for everything E2E-related and leave everything API-related to the API test suite.

Does this make sense at all? This is how I see E2E and API test suites.

Reply:
> I would avoid checking that something is not in a list, as it can quickly become a false negative.

Since we are moving product to the trash, maybe checking if the product is indeed moved to the trash (present on that trash page/tab). Just suggesting, I understand problematic but I feel like this can be automated without false negative.

Reply:
> We have separate API tests, here is [the link](https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce/tests/api-core-tests/tests?rgh-link-date=2024-01-12T13%3A32%3A18Z).
> 
> I don't see we have currently API test coverage for product deletion but there are similar scenarios like [deleting orders](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/tests/api-core-tests/tests/orders/orders.test.js?rgh-link-date=2024-01-12T13%3A32%3A18Z). I intended that it might have been better to use visual checks for everything E2E-related and leave everything API-related to the API test suite.
> 
> Does this make sense at all? This is how I see E2E and API test suites.

I know about the API tests, but I don't see why that matters in this case. My intention here was not to test the API, but to use it as a helper to check the result of an UI action. We're using the API as a helper to setup tests, why cannot we use it to perform checks?

However, in this particular case, I agree it makes sense to add extra UI checks and I did so with commit https://github.com/woocommerce/woocommerce/pull/43518/commits/c3fd577018c634f9601fc4c0b1c62cb9cd502b0e. I've added a check that the product is in the trash list, and also that it cannot be edited using a direct URL. This should be enough to validate the deletion, and so I removed the API checks, as they are now unnecessary.

Reply:
Thanks so much, the change looks good. 👍 I like how you added double-check now.
> We're using the API as a helper to setup tests, why cannot we use it to perform checks?

Using API to set up the test is fine because we're just setting up the test and we're finding the fastest way to do it, but if we use it also for test verification, we could miss any related issues produced by those actions within the UI, maybe even not related issues but connected by same action. Here is some p2 post (pcNwfB-2Rz) from another project I was working on where moving items to the trash introduced a different major issue, with the only API check, we could miss that in case they were trashed in the background.

I am however open to discussing this further as we may end up using both and not only API to verify E2E test. Please take my recommendation as well-intentioned.

Reply:
> I am however open to discussing this further as we may end up using both and not only API to verify E2E test. Please take my recommendation as well-intentioned.

Thanks @veljkho. I see thinks in the same way.